File: python/strategies/battlesnake_ultimate.py
import copy
from strategies.abstract_strategy import AbstractStrategy
from utils.spatial_utils import a_star, flood_fill
from utils.game_theory_utils import expectimax

class Strategy(AbstractStrategy):
    """The ultimate Battlesnake AI using Expectimax and a powerful heuristic."""
    SEARCH_DEPTH = 3
    HEURISTIC_WEIGHTS = {'space_adv_weight': 1.0, 'health_adv_weight': 0.1, 'food_control_weight': 5.0}

    def get_info(self):
        return {"apiversion": "1", "author": "UltimateBot", "color": "#FFFFFF", "head": "all-seeing", "tail": "ghost"}

    def on_game_move(self, game_state: dict) -> dict:
        print(f"Turn")
        
        return {"apiversion": "1", "author": "UltimateBot", "color": "#FFFFFF", "head": "all-seeing", "tail": "ghost"}
        
        # _, best_move = expectimax(
        #     game_state=game_state, depth=self.SEARCH_DEPTH, is_maximizing_player=True,
        #     evaluate_func=self._evaluate_heuristic, get_children_func=self._get_children
        # )
        # print(f"Turn {game_state['turn']}: ULTIMATE - Best move is {best_move.upper() if best_move else 'UP'}")
        # return {"move": best_move if best_move else "up"}

    def _evaluate_heuristic(self, game_state: dict) -> float:
        my_snake = game_state['you']
        if not self._is_snake_alive(my_snake['id'], game_state): return -float('inf')
        opponent = self._get_opponent(game_state)
        if opponent and not self._is_snake_alive(opponent['id'], game_state): return float('inf')
        obstacles = self._get_obstacles_from_state(game_state)
        board_width = game_state['board']['width']; board_height = game_state['board']['height']
        my_head = (my_snake['head']['x'], my_snake['head']['y'])
        my_space = flood_fill(my_head, obstacles, board_width, board_height)
        opp_space = 0
        if opponent:
            opp_head = (opponent['head']['x'], opponent['head']['y'])
            opp_space = flood_fill(opp_head, obstacles, board_width, board_height)
        space_advantage = my_space - opp_space
        health_advantage = my_snake['health'] - (opponent['health'] if opponent else 0)
        food_control_score = 0
        for food in game_state['board']['food']:
            food_coord = (food['x'], food['y'])
            my_path = a_star(my_head, food_coord, obstacles, board_width, board_height)
            opp_path = None
            if opponent: opp_path = a_star(opp_head, food_coord, obstacles, board_width, board_height)
            if my_path and (not opp_path or len(my_path) < len(opp_path)): food_control_score += 1
        return (space_advantage * self.HEURISTIC_WEIGHTS['space_adv_weight']) + \
               (health_advantage * self.HEURISTIC_WEIGHTS['health_adv_weight']) + \
               (food_control_score * self.HEURISTIC_WEIGHTS['food_control_weight'])

    def _get_children(self, game_state: dict, is_maximizing_player: bool) -> list:
        children = []
        snake_to_move = game_state['you'] if is_maximizing_player else self._get_opponent(game_state)
        if not snake_to_move or not self._is_snake_alive(snake_to_move['id'], game_state): return []
        safe_moves = self._get_safe_moves(snake_to_move, game_state)
        for move_name, next_coord in safe_moves.items():
            new_state = copy.deepcopy(game_state)
            sim_snake = self._find_snake_in_state(snake_to_move['id'], new_state)
            self._simulate_move(sim_snake, new_state, next_coord)
            children.append((move_name, new_state))
        return children

    def _simulate_move(self, snake: dict, state: dict, next_head: tuple):
        eats_food = False
        for food in state['board']['food']:
            if (food['x'], food['y']) == next_head:
                eats_food = True; state['board']['food'].remove(food); snake['health'] = 100
                break
        snake['body'].insert(0, {'x': next_head[0], 'y': next_head[1]})
        snake['head'] = {'x': next_head[0], 'y': next_head[1]}
        if not eats_food: snake['body'].pop(); snake['health'] -= 1
        snake['length'] = len(snake['body'])

    def _get_safe_moves(self, snake: dict, game_state: dict) -> dict:
        safe_moves = {}; head = (snake['head']['x'], snake['head']['y'])
        possible_moves = {"up": (0, -1), "down": (0, 1), "left": (-1, 0), "right": (1, 0)}
        for move_name, (dx, dy) in possible_moves.items():
            target_coord = (head[0] + dx, head[1] + dy)
            if not (0 <= target_coord[0] < game_state['board']['width'] and 0 <= target_coord[1] < game_state['board']['height']): continue
            is_deadly = False
            for s in game_state['board']['snakes']:
                if target_coord in {(p['x'],p['y']) for p in s['body'][:-1]}: is_deadly = True; break
            if is_deadly: continue
            for opponent in game_state['board']['snakes']:
                if opponent['id'] == snake['id']: continue
                if target_coord == (opponent['head']['x'], opponent['head']['y']):
                    if snake['length'] <= opponent['length']: is_deadly = True; break
            if is_deadly: continue
            safe_moves[move_name] = target_coord
        return safe_moves

    def _get_obstacles_from_state(self, game_state: dict) -> set:
        obstacles = set();
        for snake in game_state['board']['snakes']:
            for part in snake['body'][:-1]: obstacles.add((part['x'], part['y']))
        return obstacles

    def _get_opponent(self, game_state: dict):
        for snake in game_state['board']['snakes']:
            if snake['id'] != game_state['you']['id']: return snake
        return None

    def _find_snake_in_state(self, snake_id: str, game_state: dict):
        for snake in game_state['board']['snakes']:
            if snake['id'] == snake_id: return snake
        return None

    def _is_snake_alive(self, snake_id: str, game_state: dict):
        for snake in game_state['board']['snakes']:
            if snake['id'] == snake_id: return snake['health'] > 0
        return False

File: typescript/strategies/BattlesnakeUltimate.ts
import { AbstractStrategy } from './AbstractStrategy';
import { GameState, InfoResponse, MoveResponse } from '../types';
import { aStar, floodFill } from '../utils/spatialUtils';
import { expectimax } from '../utils/gameTheoryUtils';

interface Coord {
  x: number;
  y: number;
}

interface Snake {
  id: string;
  health: number;
  body: Coord[];
  head: Coord;
  length: number;
}

interface Board {
  width: number;
  height: number;
  food: Coord[];
  snakes: Snake[];
}

/**
 * The ultimate Battlesnake AI using Expectimax and a powerful heuristic.
 */
export class BattlesnakeUltimate extends AbstractStrategy {
  private readonly SEARCH_DEPTH = 3;
  private readonly HEURISTIC_WEIGHTS = {
    space_adv_weight: 1.0,
    health_adv_weight: 0.1,
    food_control_weight: 5.0
  };

  public get_info(): InfoResponse {
    return {
      apiversion: "1",
      author: "UltimateBot",
      color: "#FFFFFF",
      head: "all-seeing",
      tail: "ghost"
    };
  }

  public on_game_move(gameState: GameState): { move: string } {
    const [score, bestMove] = expectimax(
      gameState,
      this.SEARCH_DEPTH,
      true,
      this._evaluateHeuristic.bind(this),
      this._getChildren.bind(this)
    );
    
    const move = bestMove || 'up';
    console.log(`Turn ${gameState.turn}: ULTIMATE - Best move is ${move.toUpperCase()}`);
    return { move };
  }

  private _evaluateHeuristic(gameState: GameState): number {
    const mySnake = gameState.you;
    if (!this._isSnakeAlive(mySnake.id, gameState)) return -Infinity;
    
    const opponent = this._getOpponent(gameState);
    if (opponent && !this._isSnakeAlive(opponent.id, gameState)) return Infinity;
    
    const obstacles = this._getObstaclesFromState(gameState);
    const board: Board = gameState.board;
    const myHead = mySnake.head;
    const mySpace = floodFill(myHead, obstacles, board.width, board.height);
    
    let oppSpace = 0;
    if (opponent) {
      const oppHead = opponent.head;
      oppSpace = floodFill(oppHead, obstacles, board.width, board.height);
    }
    
    const spaceAdvantage = mySpace - oppSpace;
    const healthAdvantage = mySnake.health - (opponent?.health || 0);
    
    let foodControlScore = 0;
    for (const food of board.food) {
      const myPath = aStar(myHead, food, obstacles, board.width, board.height);
      let oppPath: Coord[] | null = null;
      
      if (opponent) {
        oppPath = aStar(opponent.head, food, obstacles, board.width, board.height);
      }
      
      if (myPath && (!oppPath || myPath.length < oppPath.length)) {
        foodControlScore += 1;
      }
    }
    
    return (
      spaceAdvantage * this.HEURISTIC_WEIGHTS.space_adv_weight +
      healthAdvantage * this.HEURISTIC_WEIGHTS.health_adv_weight +
      foodControlScore * this.HEURISTIC_WEIGHTS.food_control_weight
    );
  }

  private _getChildren(gameState: GameState, isMaximizingPlayer: boolean): [string, GameState][] {
    const children: [string, GameState][] = [];
    const snakeToMove = isMaximizingPlayer ? gameState.you : this._getOpponent(gameState);
    
    if (!snakeToMove || !this._isSnakeAlive(snakeToMove.id, gameState)) {
      return [];
    }
    
    const safeMoves = this._getSafeMoves(snakeToMove, gameState);
    
    for (const [moveName, nextCoord] of Object.entries(safeMoves)) {
      const newState = this._deepcopyGameState(gameState);
      const simSnake = this._findSnakeInState(snakeToMove.id, newState);
      if (simSnake) {
        this._simulateMove(simSnake, newState, nextCoord);
        children.push([moveName, newState]);
      }
    }
    
    return children;
  }

  private _simulateMove(snake: Snake, state: GameState, nextHead: Coord): void {
    let eatsFood = false;
    const board = state.board;
    
    // Check if snake eats food
    for (let i = 0; i < board.food.length; i++) {
      const food = board.food[i];
      if (food.x === nextHead.x && food.y === nextHead.y) {
        eatsFood = true;
        board.food.splice(i, 1);
        snake.health = 100;
        break;
      }
    }
    
    // Move snake
    snake.body.unshift({ x: nextHead.x, y: nextHead.y });
    snake.head = { x: nextHead.x, y: nextHead.y };
    
    if (!eatsFood) {
      snake.body.pop();
      snake.health -= 1;
    }
    
    snake.length = snake.body.length;
  }

  private _getSafeMoves(snake: Snake, gameState: GameState): Record<string, Coord> {
    const safeMoves: Record<string, Coord> = {};
    const head = snake.head;
    const possibleMoves: Record<string, [number, number]> = {
      up: [0, -1],
      down: [0, 1],
      left: [-1, 0],
      right: [1, 0]
    };
    
    for (const [moveName, [dx, dy]] of Object.entries(possibleMoves)) {
      const targetCoord: Coord = { x: head.x + dx, y: head.y + dy };
      
      // Check bounds
      if (
        targetCoord.x < 0 || targetCoord.x >= gameState.board.width ||
        targetCoord.y < 0 || targetCoord.y >= gameState.board.height
      ) {
        continue;
      }
      
      // Check collision with other snakes' bodies
      let isDeadly = false;
      for (const s of gameState.board.snakes) {
        for (let i = 0; i < s.body.length - 1; i++) {
          const part = s.body[i];
          if (part.x === targetCoord.x && part.y === targetCoord.y) {
            isDeadly = true;
            break;
          }
        }
        if (isDeadly) break;
      }
      if (isDeadly) continue;
      
      // Check head-to-head collision
      for (const opponent of gameState.board.snakes) {
        if (opponent.id === snake.id) continue;
        
        if (
          opponent.head.x === targetCoord.x &&
          opponent.head.y === targetCoord.y
        ) {
          if (snake.length <= opponent.length) {
            isDeadly = true;
            break;
          }
        }
      }
      if (isDeadly) continue;
      
      safeMoves[moveName] = targetCoord;
    }
    
    return safeMoves;
  }

  private _getObstaclesFromState(gameState: GameState): Set<string> {
    const obstacles = new Set<string>();
    
    for (const snake of gameState.board.snakes) {
      for (let i = 0; i < snake.body.length - 1; i++) {
        const part = snake.body[i];
        obstacles.add(`${part.x},${part.y}`);
      }
    }
    
    return obstacles;
  }

  private _getOpponent(gameState: GameState): Snake | null {
    for (const snake of gameState.board.snakes) {
      if (snake.id !== gameState.you.id) {
        return snake;
      }
    }
    return null;
  }

  private _findSnakeInState(snakeId: string, gameState: GameState): Snake | null {
    for (const snake of gameState.board.snakes) {
      if (snake.id === snakeId) {
        return snake;
      }
    }
    return null;
  }

  private _isSnakeAlive(snakeId: string, gameState: GameState): boolean {
    for (const snake of gameState.board.snakes) {
      if (snake.id === snakeId) {
        return snake.health > 0;
      }
    }
    return false;
  }

  private _deepcopyGameState(gameState: GameState): GameState {
    // Deep copy game state for simulation
    return JSON.parse(JSON.stringify(gameState));
  }
}